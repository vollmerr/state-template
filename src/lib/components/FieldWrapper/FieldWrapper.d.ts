import * as React from 'react';
import * as ReduxForm from 'redux-form';

export interface FieldWrapperProps {
  /** Label to display above input */
  label?: string;

  /** Meta information, such as errors and touched */
  meta: ReduxForm.WrappedFieldMetaProps;

  /** Help text to render under input */
  helpText?: string;

  /** Style class name to attach to button */
  className?: string;

  /** Content to render */
  children: React.ReactNode;

  /** The field should use the default required validation */
  required?: boolean;

  /** Disable the input */
  disabled?: boolean;
}

declare class FieldWrapper extends React.Component<FieldWrapperProps, {}> {}

export default FieldWrapper;
