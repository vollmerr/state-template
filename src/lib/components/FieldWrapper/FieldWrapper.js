import React from 'react';
import T from 'prop-types';
import classNames from 'classnames';

// wrapper around fields to render an optional label and help block.
// Handles displaying errors if vlaidation has failed
const FieldWrapper = (props) => {
  const {
    label,
    meta,
    name,
    helpText,
    className,
    children,
    required,
    disabled,
    ...rest
  } = props;

  const { touched, error } = meta;

  let errorMessage = '';
  if (touched && error) {
    errorMessage = error;
  }

  const cn = classNames([
    'form-group',
    className,
    { 'has-error': errorMessage },
    { disabled },
  ]);

  return (
    <div className={cn} {...rest}>
      {label && (
        <label data-test={'label'} className={'control-label'} htmlFor={name}>
          {
            required
            && !disabled
            && <span data-test={'required'} className={'required-label'}>* </span>}
          {label}
        </label>
      )}

      {children}

      {errorMessage && <small data-test={'error'} className={'feedback'}>{errorMessage}</small>}
      {helpText && <p data-test={'help'} className={'help-block'}>{helpText}</p>}
    </div>
  );
};

FieldWrapper.propTypes = {
  /** Label to display above input */
  label: T.string,

  /** Meta information, such as errors and touched */
  meta: T.object.isRequired,

  /** Name of field */
  name: T.string.isRequired,

  /** Help text to render under input */
  helpText: T.string,

  /** Style class name to attach to button */
  className: T.string,

  /** Content to render */
  children: T.element.isRequired,

  /** The field should use the default required validation */
  required: T.bool,

  /** Disable the input */
  disabled: T.bool,
};

FieldWrapper.defaultProps = {
  label: '',
  helpText: '',
  className: '',
  required: false,
  disabled: false,
};

export default FieldWrapper;
