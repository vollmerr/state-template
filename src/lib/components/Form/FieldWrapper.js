import React from 'react';
import T from 'prop-types';
import classNames from 'classnames';

// wrapper around fields to display an optional label and help block.
// Handles displaying errors if vlaidation has failed
const FieldWrapper = (props) => {
  const {
    label,
    meta,
    helpText,
    className,
    children,
    required,
  } = props;

  const { touched, error } = meta;

  let errorMessage = '';
  if (touched && error) {
    errorMessage = error;
  }

  const cn = classNames([
    'form-group',
    className,
    { 'has-error': errorMessage },
  ]);

  return (
    <div className={cn}>
      {label && (
        <label className="control-label" data-test={'label'}>
          {required && <span className="required-label" data-test={'required'}>* </span>}
          {label}
        </label>
      )}

      {children}

      {errorMessage && <small className="feedback" data-test={'error'}>{errorMessage}</small>}
      {helpText && <p className="help-block" data-test={'help'}>{helpText}</p>}
    </div>
  );
};

FieldWrapper.propTypes = {
  label: T.string,
  meta: T.object.isRequired,
  helpText: T.string,
  className: T.string,
  children: T.element.isRequired,
  required: T.bool,
};

FieldWrapper.defaultProps = {
  label: '',
  helpText: '',
  className: '',
  required: false,
};

export default FieldWrapper;
