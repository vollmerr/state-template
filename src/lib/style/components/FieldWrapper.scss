.required-label {
  color: #9F3123;
}

.field__label-tooltip {
  margin-left: 8px;

  &:hover {
    > .ca-gov-icon-question-fill {
      color: $brand-primary;
      cursor: pointer;
    }
  }
}
