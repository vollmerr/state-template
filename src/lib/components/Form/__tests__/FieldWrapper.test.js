import React from 'react';
import { shallow } from 'enzyme';

import FieldWrapper from '../FieldWrapper';

const Child = () => <input name={'test input'} />;
const props = {
  label: 'test label',
  meta: { touched: true, error: 'test error' },
  helpText: 'test help text',
  required: true,
  children: <Child />,
};

let wrapper;
describe('FieldWrapper', () => {
  beforeEach(() => {
    wrapper = shallow(<FieldWrapper {...props} />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should attach the `className` to the top level container', () => {
    wrapper.setProps({ className: 'test-class' });
    expect(wrapper.prop('className')).toMatch(/test-class/);
  });

  it('should not render a label if none provided', () => {
    wrapper.setProps({ label: '' });
    expect(wrapper.find('[data-test="label"]').length).toEqual(0);
  });

  it('should not render a required indicator if not `required`', () => {
    wrapper.setProps({ required: false });
    expect(wrapper.find('[data-test="required"]').length).toEqual(0);
  });

  it('should not render a error message if not touched or no error', () => {
    wrapper.setProps({ meta: { touched: false, error: 'test error' } });
    expect(wrapper.prop('className')).not.toMatch(/has-error/);
    expect(wrapper.find('[data-test="error"]').length).toEqual(0);
  });

  it('should not render a help text block if no `helpText`', () => {
    wrapper.setProps({ helpText: '' });
    expect(wrapper.find('[data-test="help"]').length).toEqual(0);
  });
});
