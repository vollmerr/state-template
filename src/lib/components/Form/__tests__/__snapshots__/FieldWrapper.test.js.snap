// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldWrapper should render correctly 1`] = `
<div
  className="form-group has-error"
>
  <label
    className="control-label"
    data-test="label"
  >
    <span
      className="required-label"
      data-test="required"
    >
      * 
    </span>
    test label
  </label>
  <Child />
  <small
    className="feedback"
    data-test="error"
  >
    test error
  </small>
  <p
    className="help-block"
    data-test="help"
  >
    test help text
  </p>
</div>
`;
